Content.Server/Singularity/EntitySystems/ContainmentFieldGeneratorSystem.cs

```diff
@@ -1,264 +1,387 @@
+        var ent = closestResult.Value.HitEntity;
+
+        if (!TryComp<ContainmentFieldGeneratorComponent?>(ent, out var otherFieldGeneratorComponent) ||
+            otherFieldGeneratorComponent == component ||
+            !TryComp<PhysicsComponent>(ent, out var collidableComponent) ||
+            collidableComponent.BodyType != BodyType.Static)
         {
-            var totalPower = power + component.PowerBuffer;
-            var powerPerConnection = totalPower / 2;
-            var newBuffer = totalPower % 2;
-            TryPowerConnection(ref component.Connection1, ref newBuffer, powerPerConnection, component);
-            TryPowerConnection(ref component.Connection2, ref newBuffer, powerPerConnection, component);
+            return false;
+        }
 
-            component.PowerBuffer = newBuffer;
+        var fields = GenerateFieldConnection(component, otherFieldGeneratorComponent);
+
+        component.Connections[dir] = (otherFieldGeneratorComponent, fields);
+        otherFieldGeneratorComponent.Connections[dir.GetOpposite()] = (component, fields);
+        ChangeFieldVisualizer(otherFieldGeneratorComponent);
+
+        if (!component.IsConnected)
+        {
+            component.IsConnected = true;
+            ChangeOnLightVisualizer(component);
         }
 
-        public void UpdateConnectionLights(ContainmentFieldGeneratorComponent component)
+        if (!otherFieldGeneratorComponent.IsConnected)
         {
-            if (EntityManager.TryGetComponent<PointLightComponent>(component.Owner, out var pointLightComponent))
-            {
-                bool hasAnyConnection = (component.Connection1 != null) || (component.Connection2 != null);
-                pointLightComponent.Enabled = hasAnyConnection;
-            }
+            otherFieldGeneratorComponent.IsConnected = true;
+            ChangeOnLightVisualizer(otherFieldGeneratorComponent);
         }
 
-        public void RemoveConnection(ContainmentFieldConnection? connection, ContainmentFieldGeneratorComponent component)
+        ChangeFieldVisualizer(component);
+        UpdateConnectionLights(component);
+        _popupSystem.PopupEntity(Loc.GetString("comp-containment-connected"), component.Owner, Filter.Pvs(component.Owner));
+        return true;
+    }
+
+    /// <summary>
+    /// Spawns fields between two generators if the <see cref="TryGenerateFieldConnection"/> finds two generators to connect.
+    /// </summary>
+    /// <param name="firstGenComp">The source field generator</param>
+    /// <param name="secondGenComp">The second generator that the source is connected to</param>
+    /// <returns></returns>
+    private List<EntityUid> GenerateFieldConnection(ContainmentFieldGeneratorComponent firstGenComp, ContainmentFieldGeneratorComponent secondGenComp)
+    {
+        var fieldList = new List<EntityUid>();
+        var gen1Coords = Transform(firstGenComp.Owner).Coordinates;
+        var gen2Coords = Transform(secondGenComp.Owner).Coordinates;
+
```

Comment:
How does this go if the field gens are on 2 different grids?

Reply:
I really don't know how grids work so I'd need this explained. I'd assume that this would be impossible unless it's 2 different zlevels (using the BYOND equivalent) or something.